### ISSUE TITLE
Setting the m/z value of a detected peak in MZmine v2.36 and MZmine v2.38

### ISSUE THREAD

**Issue by asmirn1** [mzmine/mzmine2#553]
Hi Tomas,

I'm not reporting a bug, but I have one suggestion.

In MZmine v2.36, the m/z value of a detected peak was the same as that of the chromatogram where that peak was detected. So, all peaks detected in one chromatogram would have the same m/z value. In MZmine v2.38, the m/z value of a detected peak is calculated individually for each peak. So, it's hard to tell which chromatogram a peak was detected in (See the attached figure).

In our work, the behavior of MZmine v2.36 is preferable, but it cannot be replicated in MZmine v2.38. I wonder, if we could add an option for parameter *m/z center calibration* in the Chromatogram Deconvolution, that would allow setting the m/z value of a detected peak to the m/z value of the corresponding chromatogram, as it was in MZmine v2.36.

I could work on adding that option.

![version_comparison](https://user-images.githubusercontent.com/13631120/55243217-216b9400-5215-11e9-9fff-132bd26ed098.png)

### Setup

MZmine v2.36 and MZmine v2.38 on MacOSX High Sierra.

**Comment by tomas-pluskal** [comment]
Hi Alex, 

I cannot support such change, because it is conceptually wrong. The m/z value of each peak should be defined by the data points of that peak, not by the process by which a chromatogram was constructed.

Perhaps if you explain what is your goal, we can find a more systematic solution.

Tomas

**Comment by asmirn1** [comment]
My goal is to check if two peaks are produced from the same chromatogram. In the attached figure, I have two peaks detected in the same chromatogram. You can see that those detected peaks are two parts of a single composite peak. When I'm doing spectral deconvolution, I get better results if I merge those two parts together. To handle such cases, I need to know when two peaks are detected in the same chromatogram. So I was using m/z values of chromatograms and detected peaks for that purpose. 

I understand that using m/z values is not the best solution. If you can think about a better one, please let me know.

![splitting](https://user-images.githubusercontent.com/13631120/55246448-9a6dea00-521b-11e9-937e-2a0f542f8066.png)

I agree that using the m/z value of a chromatogram for detected peaks may be conceptually wrong, but it was the default behavior of MZmine up until version 2.36, so maybe we could add that option for *m/z center calculation* for compatibility.

**Comment by asmirn1** [comment]
As an alternative solution, we could add a variable to class `Feature`, so that a feature would contain some sort of a reference to the feature it was made from.

**Comment by tomas-pluskal** [comment]
> I agree that using the m/z value of a chromatogram for detected peaks may be conceptually wrong, but it was the default behavior of MZmine up until version 2.36, so maybe we could add that option for m/z center calculation for compatibility.

Actually, not true. MZmine always used median m/z value, but at some point someone from your lab submitted a change to use the chromatogram m/z value, and I didn't notice until people started complaining that MZmine produces nonsense m/z values for peaks. Then we reverted the change.

**Comment by tomas-pluskal** [comment]
Yes, I agree that adding a new field to Feature is a better way to do it. But be careful that you will have to update the project saving and loading modules, too, otherwise this binding will be lost when the project is saved and re-loaded. Also, keep in mind that features can be produced by various algorithms (e.g., the targeted detector, or GridMass), so the "parent feature" value must be nullable and your module should handle the null values correctly.

**Comment by asmirn1** [comment]
Oh, I'm sorry about changing the median m/z value. I didn't know that. Usually, we try to avoid changing any original MZmine code.

What would be an appropriate identifier for Feature?

Since `PeakListRow` already has the field ID, I can add a new field to `PeakListRow` instead, so that a row would store an ID of the parent row. Would this work for you?

**Comment by robinschmid** [comment]
Hey alex, 
Hey tomas,

The expected behaviour has always been the median of the actual points. The recalculation was commented out.

You could just specify a mztolerance If you want to see whether two features were overlapping.

**Comment by tomas-pluskal** [comment]
I think it is better to add this property to Feature rather than PeakListRow, because after alignment the peak list rows contain multiple features from different files.

**Comment by tomas-pluskal** [comment]
Robin's suggestion makes sense, if you add m/z tolerance to your module, you don't need to add any new fields and refactor the project saving/loading.

**Comment by asmirn1** [comment]
I've already tried using m/z tolerance. I don't find it very reliable because:
- this parameter would depend on the m/z tolerance that was used during chromatogram building,
- using m/z tolerance may lead to merging two peaks detected in different chromatograms if those chromatograms have similar m/z values.

Thank you for your suggestion though, Robin!

Tomas, does Feature have any identifier, or I would have to implement a new Feature ID. I have a feeling that I would have to make a lot of changes in the latter case.

**Comment by tomas-pluskal** [comment]
I think you can add a method like Feature.getParentChromatogramRowID(), which will return the row ID in the chromatogram list that was used for deconvolution (or it will return null if such information is not available).

**Comment by asmirn1** [comment]
That's a good idea! I agree that each row may contain several features after alignment, but I don't think it will be a problem here. Let me think about it and try to implement it.
